lib/PuppeteerSharp/Helpers/AsyndDictionaryHelper.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System.Collections.Concurrent;
+using System.Threading.Tasks;
+
+namespace PuppeteerSharp.Helpers
+{
+    internal class AsyncDictionaryHelper<TKey, TValue>
+    {
+        public ConcurrentDictionary<TKey, TValue> Dictionary { get; }
+        public string TimeoutMessage { get; set; }
+
+        private MultiMap<TKey, TaskCompletionSource<TValue>> _pendingRequests;
+        private const int WaitForRequestDelay = 1000;
+
+        public AsyncDictionaryHelper(ConcurrentDictionary<TKey, TValue> dictionary, string timeoutMessage)
+        {
+            Dictionary = dictionary;
+            TimeoutMessage = timeoutMessage;
+            _pendingRequests = new MultiMap<TKey, TaskCompletionSource<TValue>>();
+        }
+
+        internal async Task<TValue> GetItemAsync(TKey key)
+        {
+            var tcs = new TaskCompletionSource<TValue>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _pendingRequests.Add(key, tcs);
+
+            if (Dictionary.TryGetValue(key, out var item))
+            {
+                _pendingRequests.Delete(key, tcs);
+                return item;
+            }
+
+            return await tcs.Task.WithTimeout(WaitForRequestDelay, new PuppeteerException(string.Format(TimeoutMessage, key)));
```

Comment:
maybe we could pass in a `Func<Exception>` instead of creating an exception object? less allocation.
then we could create a static method that will be passed in here